### PR TITLE
update pvector for n dimensions and add more tests

### PR DESCRIPTION
**Pull request by jeacom25b**
Enable PVector to support more than 3 dimensions using vector[index] notation and add swizzling similar to that found in GLSL.
Also adds a few more tests to be sure.

### CODE REVIEW

File: pyprocessing/math.py [hunk 1]
```diff
@@ -1,31 +1,39 @@
...
+                 'b': 2,
+                 'a': 3}
 
 
 class PVectorMeta(type):
     def __getattr__(self, attr):
         if attr == 'add':
             return self._class_add
-        if attr == 'dist':
+        elif attr == 'dist':
```
Comment (sbordeyne) [inline pyprocessing/math.py]: elifs are unnecessary.

File: pyprocessing/math.py [hunk 2]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
...
-        if z is not None:
-            self.z = z
+        self.seq = PVector(*vector, x, y, z).seq
 
     def copy(self):
-        return PVector(self.x, self.y, self._z)
+        return PVector(*self)
 
     def mag(self):
-        return math.sqrt(self.mag_sq())
```
Comment (sbordeyne) [inline pyprocessing/math.py:296]: no need for that change.

File: pyprocessing/math.py [hunk 3]
```diff
@@ -1,31 +1,39 @@
 import math
 import random
-
-from pyprocessing.calculation import lerp
-
+import re
 
 __all__ = ('PVector', )
+SWIZZLE_RE = re.compile('^[xyzwrgba]+$')
+SWIZZLE_ORDER = {'x': 0,
```
Comment (sbordeyne) [inline pyprocessing/math.py]: formatting

File: pyprocessing/math.py [hunk 4]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
...
+    def __getitem__(self, item):
+        return self.seq.__getitem__(item)
 
-    @z.setter
-    def z(self, value):
-        self._z = value
-        if value is not None:
-            self.is_2d = False
+    def __setitem__(self, item, v):
+        return self.seq.__setitem__(item, v)
```
Comment (sbordeyne) [inline pyprocessing/math.py]: self.seq[item] = v

File: pyprocessing/math.py [hunk 5]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
...
-        if self._z is None:
-            return 0
-        return self._z
+        if not all(a == b for a, b in zip(self, other)):
+            return False
+
+        return True
+
+    def __getitem__(self, item):
+        return self.seq.__getitem__(item)
```
Comment (sbordeyne) [inline pyprocessing/math.py]: self.seq[item]

File: pyprocessing/math.py [hunk 6]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
...
-        if isinstance(other, PVector):
-            return all(
-                m1 == m2 for m1, m2 in zip(self.position, other.position)
-            )
-        if isinstance(other, (list, tuple)):
-            return all(m1 == m2 for m1, m2 in zip(self.position, other))
-        raise ValueError(
-            f"Invalid operand '==' for type Pvector and {type(other)}"
-        )
+        if not isinstance(other, PVector):
```
Comment (sbordeyne) [inline pyprocessing/math.py]: PVector(0, 0, 0) != (0, 0, 0) with this implementation, which is a handy shorthand.

File: pyprocessing/math.py [hunk 7]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
...
 
     def __repr__(self):
-        return f'PVector<{", ".join(str(self).split())}> at {id(self)}'
+        return f"PVector{tuple(self.seq)}"
+
+    def __len__(self):
+        return len(self.seq)
+
+    def __iter__(self):
+        return self.seq.__iter__()
```
Comment (sbordeyne) [inline pyprocessing/math.py]: return iter(self.seq)

File: pyprocessing/math.py [hunk 8]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
...
 
     def __str__(self):
-        if self.is_2d:
-            return f'{self.x} {self.y}'
-        return f'{self.x} {self.y} {self.z}'
+        return str(tuple(self.seq))
 
     def __repr__(self):
-        return f'PVector<{", ".join(str(self).split())}> at {id(self)}'
+        return f"PVector{tuple(self.seq)}"
```
Comment (sbordeyne) [inline pyprocessing/math.py]: f"PVector<{', '.join(str(i) for i in self.seq}> at {id(self)}

File: pyprocessing/math.py [hunk 9]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
...
             return self.cross(other)
-        raise TypeError(
-            f"Invalid operand '@' for type PVector and {type(other)}"
-        )
 
     def __str__(self):
-        if self.is_2d:
-            return f'{self.x} {self.y}'
-        return f'{self.x} {self.y} {self.z}'
+        return str(tuple(self.seq))
```
Comment (sbordeyne) [inline pyprocessing/math.py]: " ".join(str(i) for i in self.seq)

File: pyprocessing/math.py [hunk 10]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
...
         raise TypeError(
             f"Invalid operand '-' for type PVector and {type(other)}"
         )
 
     def __mul__(self, other):
-        if isinstance(other, PVector):
+        if isinstance(other, (float, int)):
+            return PVector(*(a * other for a in self))
+
+        elif hasattr(other, '__iter__'):
```
Comment (sbordeyne) [inline pyprocessing/math.py]: unnecessary elif

File: pyprocessing/math.py [hunk 11]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
...
+            return PVector(*(a - b for a, b in zip(self, other)))
+
         raise TypeError(
             f"Invalid operand '-' for type PVector and {type(other)}"
         )
 
     def __mul__(self, other):
-        if isinstance(other, PVector):
+        if isinstance(other, (float, int)):
+            return PVector(*(a * other for a in self))
```
Comment (sbordeyne) [inline pyprocessing/math.py]: use a better variable name than 'a'

File: pyprocessing/math.py [hunk 12]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
...
-        self.is_2d = z is None
+    def __init__(self, *args, x=0, y=0, z=None):
+        if args:
+            self.seq = list(args)
+        else:
+            self.seq = [x, y, z] if z is not None else [x, y]
 
     def __copy__(self):
-        return self.copy()
+        return PVector(*self)
```
Comment (sbordeyne) [inline pyprocessing/math.py]: unnecessary change. First rule of programming : never repeat yourself

File: pyprocessing/math.py [hunk 13]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
         target.set(x, y)
         return target
 
-    def __init__(self, x=0, y=0, z=None):
-        self.x, self.y, self._z = x, y, z
-        self.is_2d = z is None
```
Comment (sbordeyne) [inline pyprocessing/math.py:167]: losing the is_2d shorthand is a shame, at least make it a property

File: pyprocessing/math.py [hunk 14]
```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
...
+        return target * amount + self * (1 - amount)
 
     def angle_between(self, other):
         return math.acos(self.dot(other) / (self.mag() * other.mag()))
 
     def array(self):
-        return self.position
+        return self.position()
+
+    def gen_single_component_access(letter):
```
Comment (sbordeyne) [inline pyprocessing/math.py:387]: make it private

**Review by sbordeyne — CHANGES_REQUESTED** [review]
It's pretty cool, but I'm on the fence about accepting such a major change, which is pretty far from how PVector actually behaves.

Furthermore, the flake8 commit should really have been an amend to the original commit. Please rebase and fixup your branch

File: pyprocessing/math.py (hunk 1 above)
Comment (jeacom25b) [reply pyprocessing/math.py]: elifs are a bit faster since they stop at the first match.

File: pyprocessing/math.py (hunk 14 above)
Comment (jeacom25b) [reply pyprocessing/math.py:387]: It is being deleted at the end of the execution of the scope, so gen_single_component_access basically dont exist.
It's just a quick way to avoid doing it one by one.

File: pyprocessing/math.py (hunk 11 above)
Comment (jeacom25b) [reply pyprocessing/math.py]: I renamed it to elem_a and elem_b.
its hard to think of a good name for this variable.

File: pyprocessing/math.py (hunk 1 above)
Comment (sbordeyne) [reply pyprocessing/math.py]: elifs are slower than ifs. I just benchmarked it to make sure.

```py
>>> import time
>>> def timeit(f):
...     start = time.time_ns()
...     for i in range(1000):
...             f()
...     return time.time_ns()  - start
...
>>> def elifs():
...     if var == 'a':
...             return 'a'
...     elif var == 'b':
...             return 'b'
...     elif var == 'c':
...             return 'c'
...     elif var == 'd':
...             return 'd'
...     elif var == 'e':
...             return 'e'
...     else:
...             return 'f'
...
>>> var = 'e'
>>> def ifs():
...     if var == 'a':
...             return 'a'
...     if var == 'b':
...             return b
...     if var == 'c':
...             return 'c'
...     if var == 'd':
...             return 'd'
...     if var == 'e':
...             return 'e'
...     return 'f'
...
>>> timeit(elifs)
0
>>> timeit(ifs)
501500
>>> timeit(elifs)
0
>>> timeit(elifs)
501700
>>> timeit(elifs)
501800
>>> timeit(elifs)
497900
>>> timeit(elifs)
501800
>>> timeit(elifs)
0
>>> timeit(elifs)
506700
>>> def avg(l):
...     return sum(l) / len(l)
...
>>> rel = [timeit(elifs) for i in range(100)]
>>> rel
[0, 1002900, 0, 0, 0, 1001800, 0, 0, 1003200, 0, 0, 0, 1031300, 0, 0, 1001200, 0, 0, 0, 1002900, 0, 0, 0, 1101500, 0, 529800, 0, 502000, 0, 473400, 0, 0, 0, 1003900, 0, 0, 1002600, 0, 0, 0, 1002700, 0, 0, 0, 1002600, 0, 0, 0, 1030400, 0, 0, 0, 1002900, 0, 0, 0, 1002700, 0, 0, 0, 1002600, 0, 0, 0, 1060800, 0, 502400, 0, 501200, 0, 501100, 0, 0, 1004200, 0, 0, 0, 1029800, 0, 0, 1005000, 0, 0, 0, 1000600, 0, 0, 0, 1002600, 0, 0, 0, 1002100, 0, 0, 0, 1003200, 0, 0, 0]
>>> rel = [x for x in rel if x]
>>> rel
[1002900, 1001800, 1003200, 1031300, 1001200, 1002900, 1101500, 529800, 502000, 473400, 1003900, 1002600, 1002700, 1002600, 1030400, 1002900, 1002700, 1002600, 1060800, 502400, 501200, 501100, 1004200, 1029800, 1005000, 1000600, 1002600, 1002100, 1003200]
>>> avg(rel)
907358.6206896552
>>> rif = [timeit(ifs) for i in range(100)]
>>> rif = [x for x in rif if x]
>>> avg(rif)
501788.8888888889
>>> avg(rel) < avg(rif)
False
>>> avg(rel) - avg(rif)
405569.7318007663
```

File: pyprocessing/math.py (hunk 11 above)
Comment (sbordeyne) [reply pyprocessing/math.py]: those are components of the vector

File: pyprocessing/math.py [hunk 15]
```diff
@@ -183,14 +185,14 @@ def __init__(self, *args, x=0, y=0, z=None):
...
 
     def __copy__(self):
-        return PVector(*self)
+        return self.copy()
 
     def __add__(self, other):
         if isinstance(other, complex):
             other = (other.real, other.imag)
 
         if hasattr(other, '__iter__'):
```
Comment (sbordeyne) [inline pyprocessing/math.py]: Forgot to mention it, use isinstance instead (`isinstance(other, collections.abc.Iterable)` for instance)

**Review by sbordeyne — CHANGES_REQUESTED** [review]

File: pyprocessing/math.py (hunk 1 above)
Comment (jeacom25b) [reply pyprocessing/math.py]: well, in that case.

Comment (sbordeyne) [reply pyprocessing/math.py]: just so you know, in this case, ifs are faster because the parser doesn't have to compute every case to find the first matching one (it only works because the function returns). In that test, there are 6 'blocks' of code for the parser to interpret, while in the elif case, there's only one big block. The difference is marginal (406 ms accross 1,000 calls) but it's still a bit faster, plus it looks a bit better in my opinion

Comment (jeacom25b) [reply pyprocessing/math.py]: Thats interesting, never thought about that.

File: pyprocessing/math.py (hunk 11 above)
Comment (jeacom25b) [reply pyprocessing/math.py]: Yeah, but I thought component_a + component_b is kinda long for a oneliner.

Comment (sbordeyne) [reply pyprocessing/math.py]: you can shorten it to cmpnt if need be, although component is not that long, and you can add line break (while still following the style guide obviously)

Comment (jeacom25b) [reply pyprocessing/math.py]: Ok going for line break then.

File: pyprocessing/math.py [hunk 16]
```diff
@@ -1,10 +1,20 @@
 import math
 import random
-
-from pyprocessing.calculation import lerp
-
+import re
+from collections.abc import Iterable
```
Comment (sbordeyne) [inline pyprocessing/math.py]: should be at the top of the file (alphabetical order for imports), and with 2 line breaks before the code

File: pyprocessing/math.py [hunk 17]
```diff
@@ -61,97 +72,103 @@ def _class_z():
...
             return self._instance_add
-        if attr == 'dist':
-            return self._instance_dist
-        if attr == 'sub':
+
+        elif attr == 'sub':
             return self._instance_sub
-        return super().__getattr__(attr)
+
+        elif attr == 'dist':
```
Comment (sbordeyne) [inline pyprocessing/math.py]: unnecessary elifs

File: pyprocessing/math.py [hunk 18]
```diff
@@ -61,97 +72,103 @@ def _class_z():
...
-    @classmethod
-    def set_mag(cls, mag, target=None):
-        if target is not None:
-            target.normalize()
-            target *= mag
-            return target
-        cls.normalize()
-        cls *= mag
-        return cls
+    def set_mag(self, mag, target=None):
```
Comment (sbordeyne) [inline pyprocessing/math.py:168]: Processing has a target argument for setMag, which sets the magnitude for that specific target. This behavior is lost here, please fix

File: pyprocessing/math.py [hunk 19]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
-        if isinstance(other, (list, tuple)):
-            return self.sub(*other)
-        elif isinstance(other, (PVector, complex)):
-            return self.sub(other)
+        if isinstance(other, complex):
+            other = (other.real, other.imag)
+
+        if isinstance(other, Iterable):
+            return PVector(*(component_a - component_b for
+                             component_a, component_b in zip(self, other)))
```
Comment (sbordeyne) [inline pyprocessing/math.py]: formatting

File: pyprocessing/math.py [hunk 20]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
-            f"Invalid operand '==' for type Pvector and {type(other)}"
-        )
+        if not isinstance(other, Iterable):
+            return False
+
+        if not len(other) == len(self):
+            return False
+
+        if not all(component_a == component_b for
+                   component_a, component_b in zip(self, other)):
```
Comment (sbordeyne) [inline pyprocessing/math.py]: formatting

File: pyprocessing/math.py [hunk 21]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
+
+        return True
 
     def __getitem__(self, item):
-        if isinstance(item, str):
-            return {'x': self.x, 'y': self.y, 'z': self.z}[item]
-        if isinstance(item, int):
-            return [self.x, self.y, self.z][item]
-        raise KeyError(f"Item '{item}' cannot be accessed.")
+        return self.seq[item]
```
Comment (sbordeyne) [inline pyprocessing/math.py]: Don't we lose the ability to access x, y, z, w etc by doing `self['x']` ?

If so, that's a regression of the previous code

File: pyprocessing/math.py [hunk 22]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
+
+    def __setitem__(self, item, v):
+        self.seq[item] = v
```
Comment (sbordeyne) [inline pyprocessing/math.py]: similar to `__getitem__`, allow dict-like mutating

File: pyprocessing/math.py [hunk 23]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
-                self.x, self.y, self.z = vector
-            elif isinstance(vector[0], complex):
-                self.x, self.y, self.z = vector[0].real, vector[0].imag, None
-        if x is not None:
-            self.x = x
-        if y is not None:
-            self.y = y
-        if z is not None:
-            self.z = z
+        self.seq = PVector(*vector, x, y, z).seq
```
Comment (sbordeyne) [inline pyprocessing/math.py:311]: we lose the ability to set a Vector from a complex

File: pyprocessing/math.py [hunk 24]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
     def mag(self):
         return math.sqrt(self.mag_sq())
 
     def mag_sq(self):
-        return self.x * self.x + self.y * self.y + self.z * self.z
+        return sum(component_a * component_a for component_a in self)
 
     def dot(self, other):
-        return self.x * other.x + self.y * other.y + self.z * other.z
+        return sum(component_a * component_b for component_a, component_b in zip(self, other))
```
Comment (sbordeyne) [inline pyprocessing/math.py]: I'm surprised this is not too long of a line

File: pyprocessing/math.py [hunk 25]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
 
     def cross(self, other):
-        x = self.y * other.z - self.z * other.y
-        y = self.z * other.x - self.x * other.z
-        z = self.x * other.y - self.y * other.x
+        if not len(self) == 3:
+            raise ValueError(
+                'Cross product is only defined for 3d vectors'
```
Comment (sbordeyne) [inline pyprocessing/math.py]: it's also defined in 7 dimensions by the way (it's just FYI). It may be a good idea to implement the Gram-Schmidt algorithm to find orthgonized bases in all dimensions

File: pyprocessing/math.py [hunk 26]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
+        z = self[0] * other[1] - self[1] * other[0]
         return PVector(x, y, z)
 
     def normalize(self):
-        mag = self.mag()
-        self.x /= mag
-        self.y /= mag
-        self.z /= mag
+        s = self.mag()
+        if s == 0:
```
Comment (sbordeyne) [inline pyprocessing/math.py]: why use `s` as a variable name ? `mag` or `magnitude` are more appropriate

File: pyprocessing/math.py [hunk 27]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
-        self.normalize()
-        self *= limit
+        s = math.sqrt(s) / limit
+
+        for i in range(len(self)):
+            self[i] /= s
 
     def heading(self):
-        if not self.is_2d:
-            raise ValueError('Cannot calculate heading on a 3D Vector.')
```
Comment (sbordeyne) [inline pyprocessing/math.py:325]: don't remove that. Technically, you can calculate the heading by projecting onto a plane, but since no plane is defined, we assume it's just for the XY plane, thus, only for 2D vectors. The message can be changed to reflect that

File: pyprocessing/math.py [hunk 28]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
-        if not self.is_2d:
-            raise ValueError('Cannot rotate a 3D Vector.')
-        self.x = math.cos(theta) * self.x - math.sin(theta) * self.y
-        self.y = math.sin(theta) * self.x + math.cos(theta) * self.y
+        if len(self) < 2:
+            raise ValueError('Cannot rotate a Vector'
+                             ' with less than 2 elements')
+
+        self[0] = math.cos(theta) * self[0] - math.sin(theta) * self[1]
+        self[1] = math.sin(theta) * self[0] + math.cos(theta) * self[1]
```
Comment (sbordeyne) [inline pyprocessing/math.py:367]: Rotate only works with 2D vectors. Otherwise, you have to specify an axis to rotate around.

Don't change the original meaning of the method, since that coincides with what processing already does.

On top of that, you messed up the formatting on the exception message

File: pyprocessing/math.py [hunk 29]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
         return self.position
+
+    def gen_single_component_access(letter):
+        idx = SWIZZLE_ORDER[letter]
+
+        @property
+        def getter(self):
+            if not len(self.seq) <= idx:
+                return self.seq[idx]
+            else:
```
Comment (sbordeyne) [inline pyprocessing/math.py]: unnecessary else

File: pyprocessing/math.py [hunk 30]
```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
...
+                raise IndexError(f'PVector does not have {idx+1} elements')
+
+        @getter.setter
+        def setter(self, val):
+            if not len(self.seq) <= idx:
+                self.seq[idx] = val
+            else:
```
Comment (sbordeyne) [inline pyprocessing/math.py]: unnecessary else (add a return above)

**Review by sbordeyne — CHANGES_REQUESTED** [review]
I caught a few more things by reviewing. Please fix them (and amend your original commit with those changes)

Since it's such a big PR, it's no surprise (plus I'm the only one reviewing right now, please forgive the incessant back and forth)

File: pyprocessing/math.py (hunk 25 above)
Comment (jeacom25b) [reply pyprocessing/math.py]: Really?, I didn't know that, well, thats a little too advanced for my current knowledge of linear algebra.
I will just change the exception text for now since it will take me at least a month to learn it, if I can ever learn.

File: pyprocessing/math.py (hunk 27 above)
Comment (jeacom25b) [reply pyprocessing/math.py:325]: Sure, I forget we can do `vector.xy.heading()`

File: pyprocessing/math.py (hunk 28 above)
Comment (jeacom25b) [reply pyprocessing/math.py:367]: I added a new method called rotated() which returns a new vector, so we can do `vec.yz = vec.yz.rotated(theta)` to pick which plane to rotate about.

File: pyprocessing/math.py (hunk 23 above)
Comment (jeacom25b) [reply pyprocessing/math.py:311]: Now \_\_init__ supports complex.

File: pyprocessing/math.py [hunk 31]
```diff
@@ -61,97 +73,105 @@ def _class_z():
...
-        return vector
+        return cls.random_n(3)
+
+    # generic swizzling methods
+    def __swizzle_get(self, components):
+        try:
+            if len(components) == 1:
+                return self[SWIZZLE_ORDER[components]]
+
+            return PVector(*(self[SWIZZLE_ORDER[letter]]
```
Comment (sbordeyne) [inline pyprocessing/math.py]: formatting

File: pyprocessing/math.py [hunk 32]
```diff
@@ -61,97 +73,105 @@ def _class_z():
...
+    def __swizzle_set(self, components, values):
+        if isinstance(values, (float, int)):
+            values = (values,) * len(components)
+
+        elif isinstance(values, complex):
+            values = (values.real, values.imag)
+
+        if not len(components) == len(values):
+            raise ValueError(
+                'sequence is not of the same size as the swizzle')
```
Comment (sbordeyne) [inline pyprocessing/math.py]: formatting of the last bracket

File: pyprocessing/math.py [hunk 33]
```diff
@@ -162,191 +182,236 @@ def from_angle(angle, target=None):
...
             raise ValueError('Cannot calculate heading on a 3D Vector.')
-        return math.atan2(self.y, self.x)
+        return math.atan2(self[1], self[0])
 
     def rotate(self, theta):
         if not self.is_2d:
-            raise ValueError('Cannot rotate a 3D Vector.')
-        self.x = math.cos(theta) * self.x - math.sin(theta) * self.y
-        self.y = math.sin(theta) * self.x + math.cos(theta) * self.y
+            raise ValueError(f'Can\'t rotate a non 2D vector')
```
Comment (sbordeyne) [inline pyprocessing/math.py]: use double quotes (`"`) instead of escaping

**Review by sbordeyne — CHANGES_REQUESTED** [review]
Just these last changes left

**Comment by sbordeyne** [comment]
LGTM, merged
